src/permutations.rs

```diff
@@ -81,12 +81,9 @@ where
             }
             PermutationState::Buffered { ref k, min_n } => {
                 if vals.get_next() {
-                    let item = (0..*k - 1)
-                        .chain(once(*min_n))
-                        .map(|i| vals[i].clone())
-                        .collect();
+                    let indices: Vec<usize> = (0..k - 1).chain(once(*min_n)).collect();
                     *min_n += 1;
-                    Some(item)
+                    Some(vals.get_at(&indices))
```

Comment:
We strive to make as few heap allocations as possible. This creates an unnecessary vector `indices` so I would change this back. `get_at` is useful but has its limitations.

I've been thinking that it could be `fn get_at(indices: impl Iterator<Item = usize>) ...` instead but I prefer to keep this as simple as possible (and it would requires to change `&indices` to `indices.iter().copied()` which would kinda beat the purpose of this helper, or `impl IntoIterator<Item = &usize>` but it starts to be weird).

Reply:
Yeah I was questioning this one myself. I'll push a patch shortly